internal/venice-common/src/main/java/com/linkedin/venice/writer/CompositeVeniceWriter.java

```diff
@@ -0,0 +1,182 @@
+
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.pubsub.api.PubSubProduceResult;
+import com.linkedin.venice.pubsub.api.PubSubProducerCallback;
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+
+
+/**
+ * The composite writer contains a main writer and multiple child writers. The main writer will only perform the write
+ * once all of its child writers are complete.
+ * TODO The child writers are view writers. Ideally to avoid code duplication we should be using an array of
+ * VeniceViewWriter here. However, the current implementation of VeniceViewWriter involves PCS which is something
+ * specific to the ingestion path that we don't want to leak into venice-common.
+ */
+public class CompositeVeniceWriter<K, V, U> extends AbstractVeniceWriter<K, V, U> {
+  private final VeniceWriter<K, V, U> mainWriter;
+  private final VeniceWriter<K, V, U>[] childWriters;
+  private final PubSubProducerCallback childCallback;
+
+  private CompletableFuture<PubSubProduceResult> lastWriteFuture = CompletableFuture.completedFuture(null);
+
+  public CompositeVeniceWriter(
+      String topicName,
+      VeniceWriter<K, V, U> mainWriter,
+      VeniceWriter<K, V, U>[] childWriters,
+      PubSubProducerCallback childCallback) {
+    super(topicName);
+    this.mainWriter = mainWriter;
+    this.childWriters = childWriters;
+    this.childCallback = childCallback;
+  }
+
+  @Override
+  public void close(boolean gracefulClose) throws IOException {
+    for (VeniceWriter veniceWriter: childWriters) {
+      veniceWriter.close(gracefulClose);
+    }
+    mainWriter.close(gracefulClose);
+  }
+
+  @Override
+  public CompletableFuture<PubSubProduceResult> put(
+      K key,
+      V value,
+      int valueSchemaId,
+      PubSubProducerCallback callback) {
+    return put(key, value, valueSchemaId, callback, null, false);
+  }
+
+  @Override
+  public Future<PubSubProduceResult> put(
+      K key,
+      V value,
+      int valueSchemaId,
+      PubSubProducerCallback callback,
+      PutMetadata putMetadata) {
+    return put(key, value, valueSchemaId, callback, putMetadata, true);
+  }
+
+  CompletableFuture<PubSubProduceResult> put(
+      K key,
+      V value,
+      int valueSchemaId,
+      PubSubProducerCallback callback,
+      PutMetadata putMetadata,
+      boolean hasPutMetadata) {
+    CompletableFuture<PubSubProduceResult> finalFuture = new CompletableFuture<>();
+    CompletableFuture[] childFutures = new CompletableFuture[childWriters.length + 1];
+    int index = 0;
+    childFutures[index++] = lastWriteFuture;
+    for (VeniceWriter<K, V, U> writer: childWriters) {
+      childFutures[index++] = hasPutMetadata
+          ? writer.put(
+              key,
+              value,
+              valueSchemaId,
+              childCallback,
+              DEFAULT_LEADER_METADATA_WRAPPER,
+              APP_DEFAULT_LOGICAL_TS,
+              putMetadata)
```

Comment:
One small perf enhancement:
View Topics doesn't really need a full RMD, and maybe we only need to keep `rmd offset` for changelog consumer and the tombstone values are useless as there won't be any leader replicas for View Topics.

Reply:
Correct me if I'm wrong but this RMD record is currently passed around as a `ByteBuffer` which means we need to deserialize in order to extract and keep only the rmd offset. Can we decide on these performance optimizations later once we see some performance metrics on composite writer and view writers?